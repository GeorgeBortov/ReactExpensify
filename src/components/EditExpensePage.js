import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import ConfirmModal from './ConfirmModal';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';
import { openModal, closeModal } from '../actions/modal';

export class EditExpensePage extends React.Component {
    openModal = () => {
        this.props.openModal();
    }
    closeModal= () => {
        this.props.closeModal();
    }
    onSubmit = (expense) => {
        this.props.startEditExpense(this.props.expense.id, expense);
        this.props.history.push('/');
    }
    onRemove = () => {
        this.props.startRemoveExpense({id: this.props.expense.id});
        this.props.closeModal();
        this.props.history.push('/');
    }
    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Edit Expense</h1>
                    </div>
                </div>
                {
                    this.props.expense ? (
                        <div className="content-container">
                            <ExpenseForm
                                expense={this.props.expense}
                                onSubmit={this.onSubmit}
                            />
                            <button className="button button--secondary" onClick={this.openModal}>Remove Expense</button>
                            <ConfirmModal
                                isOpen={this.props.modalStatus}
                                onRequestClose={this.closeModal}
                                expenseDescription={this.props.expense.description}
                                onRemove={this.onRemove}
                            />
                        </div>
                    ) : (
                        <div className="content-container">
                            <p className="message">
                                This expense dose not exist
                            </p>
                        </div>
                    )
                }
                
            </div>
        );
    }
};

const mapStateToProps = (state, props) => ({  
    expense: state.expenses.find((expense) => expense.id === props.match.params.id),
    modalStatus: state.modal.modalIsOpen,
    
});

const mapDispatchToProps = (dispatch, props) => ({
    startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
    startRemoveExpense: (data) => dispatch(startRemoveExpense(data)),
    openModal: () => dispatch(openModal()),
    closeModal: () => dispatch(closeModal()),
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);