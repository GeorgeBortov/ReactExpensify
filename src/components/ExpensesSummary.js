import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import numeral from 'numeral';
import selectVisibleExpenses from '../selectors/expenses-visible';
import selectHiddenExpenses from '../selectors/expenses-hidden';
import selectExpensesTotal from '../selectors/expenses-total';

export const ExpensesSummary = ({ visibleExpenseCount, visibleExpensesTotal, hiddenExpenseCount, hiddenExpensesTotal}) => {
    const viewingExpenseWord = visibleExpenseCount === 1 ? 'expense' : 'expenses';
    const visibleFormatedExpensesTotal = numeral(visibleExpensesTotal / 100).format('$0,0.00');
    const hiddenExpenseWord = visibleExpenseCount === 1 ? 'expense' : 'expenses';
    const hiddenFormatedExpensesTotal = numeral(hiddenExpensesTotal / 100).format('$0,0.00');
    const totalSum = numeral((visibleExpensesTotal + hiddenExpensesTotal) / 100).format('$0,0.00');
    return (
        <div className="page-header">
            <div className="content-container">
                <h1 className="page-header__title">Viewing <span>{visibleExpenseCount}</span> {viewingExpenseWord} totalling <span>{visibleFormatedExpensesTotal}</span></h1>
                <h2 className="page-header__title">Hidden <span>{hiddenExpenseCount}</span> {hiddenExpenseWord} totalling <span>{hiddenFormatedExpensesTotal}</span></h2>
                <h3 className="page-header__title">Total sum <span>{totalSum}</span></h3>
                <div className="page-header__actions">
                    <Link className="button" to="/create">Add Expense</Link>
                </div>
            </div>
        </div>
    );
};

const mapStateToProps = (state) => {
    const visibleExpenses = selectVisibleExpenses(state.expenses, state.filters);
    const hiddenExpenses = selectHiddenExpenses(state.expenses, visibleExpenses);
    return {
        visibleExpenseCount: visibleExpenses.length,
        hiddenExpenseCount: hiddenExpenses.length,
        visibleExpensesTotal: selectExpensesTotal(visibleExpenses),
        hiddenExpensesTotal: selectExpensesTotal(hiddenExpenses),
    };
};

export default connect(mapStateToProps)(ExpensesSummary);