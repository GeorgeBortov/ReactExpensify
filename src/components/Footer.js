import React from 'react';
import { Link } from 'react-router-dom';

export const Footer = ({ startLogout }) => (
    <header className="footer">
        <div className="content-container">
            <div className="footer__content">
                <Link className="header_link" to="/">
                   Home
                </Link>
                <Link className="header_link" to="/privacy-policy">
                    Privacy Policy
                </Link>
            </div>
        </div>
    </header>
);

export default Footer;