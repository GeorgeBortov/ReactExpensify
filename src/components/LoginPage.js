import React from 'react';
import { connect } from 'react-redux';
import { firebase } from '../firebase/firebase';
import {
    startGoogleLogin,
    startFacebookLogin,
    startGithubLogin,
    startTwitterLogin
} from '../actions/auth';
import ConftrmAuthModal from './ConftrmAuthModal';
import { openModal, closeModal } from '../actions/modal';
export class LoginPage extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            credential: props.credential ? props.credential : '',
            currProvider: props.currProvider ? props.currProvider : ''
        }
    }
    openModal = () => {
        this.props.openModal();
    }
    closeModal= () => {
        this.props.closeModal();
    }
    getProviderForProviderId = (name) => {
        switch (name) {
            case 'google.com':
                return new firebase.auth.GoogleAuthProvider();
            case 'facebook.com':
                return new firebase.auth.FacebookAuthProvider();
            case 'github.com':
                return new firebase.auth.GithubAuthProvider();
            case 'twitter.com':
                return new firebase.auth.TwitterAuthProvider();
            default:
                return undefined;
        }
    }
    addNewAccount = () => {
        this.props.closeModal();
        firebase.auth().signInWithPopup(this.state.currProvider).then((result) => {
            result.user.linkAndRetrieveDataWithCredential(this.state.credential);
        });
    }
    auth = (provider) => {
        provider().catch(({ code, credential, email }) => {
            if (code === "auth/account-exists-with-different-credential") {
                this.setState(() => ({ credential }));
                firebase.auth().fetchSignInMethodsForEmail(email).then((methods) => {
                    const currProvider = this.getProviderForProviderId(methods[0]);
                    this.setState(() => ({ existAccount: methods[0] }));
                    this.setState(() => ({ currProvider }));
                    this.openModal();
                });                
            } else {
                console.log('Error: ', error);
            }
        });
    }
    render() {
        return (
            <div className="box-layout">
                <div className="box-layout__box">
                    <h1 className="box-layout__title">Expensify</h1>
                    <p>It's time to get your expenses under control.</p>
                    <button className="button button--google-login" onClick={() => this.auth(this.props.startGoogleLogin)}>
                        <span className="button--soc-icon">
                            <img src="/images/google-ico.png" />
                        </span>
                        <span>
                            Login with Google
                        </span>
                    </button>
                    <button className="button button--facebook-login" onClick={() => this.auth(this.props.startFacebookLogin)}>
                        <span className="button--soc-icon">
                            <img src="/images/facebook-ico.png" />
                        </span>
                        <span>
                            Login with Facebook
                        </span>
                    </button>
                    <button className="button button--github-login" onClick={() => this.auth(this.props.startGithubLogin)}>
                        <span className="button--soc-icon">
                            <img src="/images/github-ico.png" />
                        </span>
                        <span>
                            Login with Github
                        </span>
                    </button>
                    <button className="button button--twitter-login" onClick={() => this.auth(this.props.startTwitterLogin)}>
                        <span className="button--soc-icon">
                            <img src="/images/twitter-ico.png" />
                        </span>
                        <span>
                            Login with Twitter
                        </span>
                    </button>
                </div>
                <ConftrmAuthModal
                    isOpen={this.props.modalStatus}
                    onRequestClose={this.closeModal}
                    onAuth={this.addNewAccount}
                    existAccount={this.state.existAccount}
                />
            </div>
        );
    }
};

const mapDispatchToProps = (dispatch) => ({
    startGoogleLogin: () => dispatch(startGoogleLogin()),
    startFacebookLogin: () => dispatch(startFacebookLogin()),
    startGithubLogin: () => dispatch(startGithubLogin()),
    startTwitterLogin: () => dispatch(startTwitterLogin()),
    openModal: () => dispatch(openModal()),
    closeModal: () => dispatch(closeModal())
});

const mapStateToProps = (state, props) => ({  
    modalStatus: state.modal.modalIsOpen
});


export default connect(mapStateToProps, mapDispatchToProps)(LoginPage);