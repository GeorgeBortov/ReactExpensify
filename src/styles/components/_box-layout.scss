.box-layout {
    background: url('/images/bg.jpg') no-repeat center / cover;
    display: -ms-flexbox;
    display: -webkit-flex;
    display: flex;
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
    -webkit-flex-wrap: nowrap;
    -ms-flex-wrap: nowrap;
    flex-wrap: nowrap;
    -webkit-justify-content: center;
    -ms-flex-pack: center;
    justify-content: center;
    -webkit-align-content: stretch;
    -ms-flex-line-pack: stretch;
    align-content: stretch;
    -webkit-align-items: center;
    -ms-flex-align: center;
    align-items: center;
    min-height: 100vh;
    position: relative;
    width: 100%;
    @media (min-width: $desktop-breakpoint) {
        display: block;
    }
    &__box {
        background: fade-out(#FFF, .15);
        @include border-radius(3px);
        @include transform(translate(none none));
        left: auto;
        top: auto;
        position: relative;
        margin: 1.2rem 0;
        padding: 1.2rem $m-size;
        text-align: center;
        width: 30rem;
        @media (min-width: $desktop-breakpoint) {
            margin: 0;
            padding: $l-size $m-size;
            @include transform(translate(-50%, -50%));
            left: 50%;
            top: 50%;
            position: absolute;
        }
    }
    &__title {
        font-size: 2.2em;
        margin: 0 0 $m-size 0;
        line-height: 1;
    }
    .button {
        @include border-radius(4px);
        font-size: 2rem;
        margin-bottom: 10px;
        padding: 0;
        text-align: left;
        width: 100%;
        span {
            display: inline-block;
            padding: 0.8rem;
            vertical-align: middle;
        }
        &--soc-icon {
            border-right: 1px solid rgba(0, 0, 0, .2);
            padding: 0.8rem 0!important;
            margin-right: 0.3rem;
            text-align: center;
            width: 42px;
        }
        &--google-login {
            background: rgb(223, 74, 50);
            border: 1px solid darken(rgb(223, 74, 50), 15%);
            &:hover {
                background: darken(rgb(223, 74, 50), 5%);
                border-color: darken(rgb(223, 74, 50), 20%);
            }
        }
        &--facebook-login {
            background: rgb(57, 87, 154);
            border: 1px solid darken(rgb(57, 87, 154), 15%);
            &:hover {
                background: darken(rgb(57, 87, 154), 5%);
                border-color: darken(rgb(57, 87, 154), 20%);
            }
        }
        &--github-login {
            background: rgb(68, 68, 68);
            border: 1px solid darken(rgb(68, 68, 68), 15%);
            &:hover {
                background: darken(rgb(68, 68, 68), 5%);
                border-color: darken(rgb(68, 68, 68), 20%);
            }
        }
        &--twitter-login {
            background: rgb(80, 171, 241);
            border: 1px solid darken(rgb(80, 171, 241), 15%);
            margin-bottom: 0;
            &:hover {
                background: darken(rgb(80, 171, 241), 5%);
                border-color: darken(rgb(80, 171, 241), 20%);
            }
        }
    }
}