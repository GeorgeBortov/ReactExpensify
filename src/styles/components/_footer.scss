.footer {
    background: $dark-blue;
    position:fixed;
    left:0;
    bottom:0;
    right:0;
    &__content {
        display: flex;
        justify-content: center;
        align-items: center;
    }
    
}