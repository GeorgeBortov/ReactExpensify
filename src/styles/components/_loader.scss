.loader {
    align-items: center;
    display: flex;
    height: 93vh;
    height: calc(100vh - 45px);
    justify-content: center;
    width: 100%;
    &__image {
        height: 6rem;
        width: 6rem;
    }
}