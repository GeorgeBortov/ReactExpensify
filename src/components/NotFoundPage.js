import React from 'react';
import Footer from './Footer';
const NotFoundPage = () => (
    <div>
        <div className="content-container">
            404 - Page Not Found
        </div>
        <Footer />
    </div>
);

export default NotFoundPage;