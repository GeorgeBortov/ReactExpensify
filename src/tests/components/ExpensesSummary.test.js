import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import selectExpensesTotal from '../../selectors/expenses-total';
import expenses from '../fixtures/expenses';

test('should correctly render ExpensesSummary with 1 hidden expense', () => {
    const total = selectExpensesTotal([expenses[0]]);

    const wrapper = shallow(<ExpensesSummary 
        visibleExpenseCount={0} 
        visibleExpensesTotal={0}
        hiddenExpenseCount={1}
        hiddenExpensesTotal={total}
    />);
    expect(wrapper).toMatchSnapshot();
});

test('should correctly render ExpensesSummary with 1 visible expense', () => {
    const total = selectExpensesTotal([expenses[2]]);
    
    const wrapper = shallow(<ExpensesSummary 
        visibleExpenseCount={1} 
        visibleExpensesTotal={total}
        hiddenExpenseCount={0}
        hiddenExpensesTotal={0}
    />);
    expect(wrapper).toMatchSnapshot();
});

test('should correctly render ExpensesSummary with multiple expenses', () => {
    const visible = [ expenses[0], expenses[2] ];
    const hidden = [expenses[1]]
    const totalVisible = selectExpensesTotal(visible);
    const totalHidden = selectExpensesTotal(hidden);

    const wrapper = shallow(<ExpensesSummary 
        visibleExpenseCount={visible.length}
        visibleExpensesTotal={totalVisible}
        hiddenExpenseCount={hidden.length}
        hiddenExpensesTotal={totalHidden}
    />);
    expect(wrapper).toMatchSnapshot();
});
