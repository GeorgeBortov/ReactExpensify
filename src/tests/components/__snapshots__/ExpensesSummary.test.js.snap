// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render ExpensesSummary with 1 hidden expense 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        0
      </span>
       
      expenses
       totalling 
      <span>
        $0.00
      </span>
    </h1>
    <h2
      className="page-header__title"
    >
      Hidden 
      <span>
        1
      </span>
       
      expenses
       totalling 
      <span>
        $1.95
      </span>
    </h2>
    <h3
      className="page-header__title"
    >
      Total sum 
      <span>
        $1.95
      </span>
    </h3>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should correctly render ExpensesSummary with 1 visible expense 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        1
      </span>
       
      expense
       totalling 
      <span>
        $45.00
      </span>
    </h1>
    <h2
      className="page-header__title"
    >
      Hidden 
      <span>
        0
      </span>
       
      expense
       totalling 
      <span>
        $0.00
      </span>
    </h2>
    <h3
      className="page-header__title"
    >
      Total sum 
      <span>
        $45.00
      </span>
    </h3>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should correctly render ExpensesSummary with multiple expenses 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        2
      </span>
       
      expenses
       totalling 
      <span>
        $46.95
      </span>
    </h1>
    <h2
      className="page-header__title"
    >
      Hidden 
      <span>
        1
      </span>
       
      expenses
       totalling 
      <span>
        $1,095.00
      </span>
    </h2>
    <h3
      className="page-header__title"
    >
      Total sum 
      <span>
        $1,141.95
      </span>
    </h3>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;
